### PR TITLE
Leaves - Morgan

### PR DESCRIPTION
**Pull request by morganschuler**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | An ArgumentError is raised when a method has been called that has the incorrect number of arguments. This alerts the user and often breaks the program.  
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? | The .all class methods exposes all of the data to the rest of our application, whereas in a class variable it would not be able to be accessed outside of the class. With an instance method we would just be looking at one customer or instance, when we really need to be looking at the class as a whole. Similarly with .find is looking through the entire data collection to find something, rather than just one instance of it. 
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | The relationship is one to many. One customer could have several different orders, however each order is always going to have one person who "owns" that order. It compares in that the solar system could have many different planets, but each planet could only have one solar system. 
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | Their relationship is tracked through an "id" number that is used throughout the program. My order class had to require the customer file to create that connection.
Did the presence of automated tests change the way you thought about the problem? How? | Yes, having the automated tests changed how I thought about the problem because I had to consider the edge cases that the tests might be looking for. It's nice that the test cases give you information about that specifically failed in each test (albeit a little worrying when you get a lot of red). I also had to make sure that I was naming all of my methods correctly - this was a problem I had in the beginning and made many of my tests fail.

### CODE REVIEW

File: test/order_test.rb [hunk 1]
```diff
@@ -60,66 +60,66 @@
...
       expect(order.total).must_equal before_total
     end
   end
 end
 
 # TODO: change 'xdescribe' to 'describe' to run these tests
-xdescribe "Order Wave 2" do
+describe "Order Wave 2" do
   describe "Order.all" do
     it "Returns an array of all orders" do
```
Comment (beccaelenzil) [inline test/order_test.rb:119]: You missed the implementation of this test.

File: test/order_test.rb [hunk 2]
```diff
@@ -129,33 +129,39 @@
...
       # Check that all data was loaded as expected
       expect(order.id).must_equal id
       expect(order.products).must_equal products
       expect(order.customer).must_be_kind_of Customer
       expect(order.customer.id).must_equal customer_id
       expect(order.fulfillment_status).must_equal fulfillment_status
     end
-
+    
     it "Returns accurate information about the last order" do
```
Comment (beccaelenzil) [inline test/order_test.rb:143]: You're missing the implementation for this test.

File: lib/order.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+class Order
+  attr_reader :id 
+  attr_accessor :products, :customer, :fulfillment_status
+  
+  def initialize(id, products, customer, fulfillment_status = :pending)
+    @id = id
+    @products = products
+    @customer = customer
+    @fulfillment_status = fulfillment_status
+    until[:paid, :complete, :processing, :shipped, :pending].include?(fulfillment_status)
```
Comment (beccaelenzil) [inline lib/order.rb:13]: This list of statuses would be a great thing to store in a constant.

File: lib/customer.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        :city => line[3], 
+        :state => line[4], 
+        :zip => line[5]}
+        @customers << Customer.new(id.to_i, email, address)
+      end
+      return @customers    
+    end
+    
+    def self.find(id)
+      customers = Customer.all
```
Comment (beccaelenzil) [inline lib/customer.rb:29]: Instead of saying `Customer.all` here, you could say just `all`.

File: lib/order.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      products = product_hash
+      customer = Customer.find(line[2].to_i)   
+      fulfillment_status = line[3].to_sym
+      orders = Order.new(id, products, customer, fulfillment_status)
+      new_order << orders
+    end
+    return new_order
+  end
+  
+  def self.find(id)
```
Comment (beccaelenzil) [inline lib/order.rb:51]: This looks good! In case you want something shorter that does exactly what you implemented, you might want to check out: https://ruby-doc.org/core-2.6.1/Enumerable.html#method-i-find

**Comment by beccaelenzil** [comment]
# Grocery Store
## What We're Looking For



Feature | Feedback
--- | ---
**Baseline** |
Answered comprehension questions | check
Used Git Regularly | check
**Wave 1** |
All provided tests pass | check
Using the appropriate `attr_` for instance variables | check
**Wave 2** |
All stubbed tests are implemented fully and pass | check
Used `CSV` library only in `.all` (not in `.find`) | you're missing a few
Appropriately parses the product data from CSV file in `Order.all` | check
`Order.all` calls `Customer.find` to set up the composition relation | check
**Additional Notes** | Good job overall! Your code is well-organized and easy to read, and does a good job of solving the problem at hand. I've left a few inline comments for you to review, but it's clear that the learning goals for this assignment were met. Keep up the hard work!

**Comment by beccaelenzil** [comment]
Feature | Feedback
--- | ---
**Baseline** |
Answered Comprehensions Questions | 1. check -- in what situation does it not "break" the program? 2. check, 3. check -- yes, Customer to Order is one to many and Order to Customer is many to one., 4. yes, composition!
Used Git Regularly | check
**Wave 1** |
All provided tests pass | check
Using the appropriate `attr_` for instance variables | check
**Wave 2** |
All stubbed tests are implemented fully and pass | check
Used `CSV` library only in `.all` (not in `.find`) | you're missing a few
Appropriately parses the product data from CSV file in `Order.all` | check
`Order.all` calls `Customer.find` to set up the composition relation | check
**Additional Notes** | Good job overall! Your code is well-organized and easy to read, and does a good job of solving the problem at hand. I've left a few inline comments for you to review, but it's clear that the learning goals for this assignment were met. Keep up the hard work!
